### PR TITLE
fix: cache oidc issuer provider

### PR DESCRIPTION
**Pull request by bcho**

### CODE REVIEW

File: auth/providers/azure/azure.go [hunk 1]
```diff
@@ -80,11 +80,48 @@ type authInfo struct {
...
+// TODO: combine auth info and issuer into one
 var (
 	cachedAuthInfo *authInfo
 	mutex          = &sync.Mutex{}
+
+	cachedOIDCIssuerProviders map[string]*oidc.Provider = make(map[string]*oidc.Provider)
+	cachedOIDCIssuerProvidersMutex = &sync.RWMutex{}
 )
 
+func getOIDCIssuerProvider(issuerURL string, issuerGetRetryCount int) (*oidc.Provider, error) {
```
Comment (weinong) [inline auth/providers/azure/azure.go:92]: this issuer really wont change

Comment (bcho) [reply auth/providers/azure/azure.go:102]: updated, ptal

File: auth/providers/azure/azure.go [hunk 2]
```diff
@@ -80,11 +80,55 @@ type authInfo struct {
...
+	if cached, ok := getCachedOIDCIssuerProvider(); ok {
+		return cached, nil
+	}
+
+	// slow path: construct from remote
+	// NOTE: we hold the lock even it's doing HTTP call to avoid sending multiple requests
+	cachedOIDCIssuerProvidersMutex.Lock()
+	defer cachedOIDCIssuerProvidersMutex.Unlock()
+
+	if cached, ok := getCachedOIDCIssuerProvider(); ok {
```
Comment (jwilder) [inline auth/providers/azure/azure.go:113]: This will deadlock.  It tries to take an RLock while a Lock is already held.

**Review by jwilder — CHANGES_REQUESTED** [review]

File: auth/providers/azure/azure.go (hunk 2 above)
Comment (bcho) [reply auth/providers/azure/azure.go:113]: yes... sorry about the rush, writing a test in local

File: auth/providers/azure/azure.go [hunk 3]
```diff
@@ -80,11 +80,58 @@ type authInfo struct {
...
+
+	if cached, ok := getCachedOIDCIssuerProviderUnsafe(); ok {
+		// another goroutine has already constructed the provider
+		return cached, nil
+	}
+
+	// NOTE: we start a root context here to allow background remote key set refresh
+	ctx := context.Background()
+	ctx = withRetryableHttpClient(ctx, issuerGetRetryCount)
+	fmt.Println(issuerURL)
```
Comment (weinong) [inline auth/providers/azure/azure.go]: debug? use klog instead

File: auth/providers/azure/azure.go [hunk 4]
```diff
@@ -80,11 +80,58 @@ type authInfo struct {
...
```
Comment (jwilder) [inline auth/providers/azure/azure.go]: Remove?

Comment (bcho) [reply auth/providers/azure/azure.go]: removed!

**Review by jwilder — APPROVED** [review]

**Comment by bcho** [comment]
I also added a simple test case to simulate the concurrent access, tested with `-race` from local:

```
go test -v -race ./auth/providers/azure
```

All tests passed.

**Review by weinong — APPROVED** [review]
LGTM, thanks!

**Review by weinong — APPROVED** [review]
LGTM

**Pull request merged**